File: src/AD.InMemoryStore.Tests/VersionTests.cs
﻿namespace AD.InMemoryStore.Tests;

[TestClass]
public sealed class VersionTests
{
    readonly KeyValueStore<int, string> store = new();

    [TestMethod("To string and parse round trip")]
    public void ToStringParse()
    {
        var expected = store.Add(1, "A");
        var actual = Version.Parse(expected.ToString());
        Assert.AreEqual(expected, actual);
    }

    [DataTestMethod()]
    [DataRow(null)]
    [DataRow("")]
    [DataRow("1")]
    [DataRow("\"1")]
    [DataRow("1\"")]
    [DataRow("a")]
    [DataRow("\"a\"")]
    [DataRow("\"-1\"")]
    [ExpectedException(typeof(ArgumentException))]
    public void InvalidVersionParse(string text)
    {
        Version.Parse(text);
    }
}

File: src/AD.InMemoryStore.Functional.Tests/VersionTests.fs
namespace AD.InMemoryStore.Functional.Tests

open Microsoft.VisualStudio.TestTools.UnitTesting
open AD.InMemoryStore.Functional

[<TestClass>]
type VersionTests () =
    
    let store = KeyValueStore ()

    [<TestMethod>]
    member _.``To string and parse round trip`` () =
        match (1, "A") |> store.Add with
        | Error _ -> Assert.Fail "Ok expected"
        | Ok (_, _, expected) ->
            let actual = expected.ToString() |> Version.parse
            Assert.AreEqual<_> (Some expected, actual)
